src/gui/folderwatcher_linux.cpp

```diff
@@ -88,41 +64,47 @@ void FolderWatcherPrivate::inotifyRegisterPath(const QString& path)
 
 void FolderWatcherPrivate::slotAddFolderRecursive(const QString &path)
 {
-    int subdirs = 0;
     qDebug() << "(+) Watcher:" << path;
-
-    QDir inPath(path);
-    inotifyRegisterPath(inPath.absolutePath());
-
     const QStringList watchedFolders = _watches.values();
+    int subdirs = addFolderRecursiveHelper(path, watchedFolders.toSet());
+    if (subdirs >0) {
+        qDebug() << "    `-> and" << subdirs << "subdirectories";
+    }
+}
 
-    QStringList allSubfolders;
-    if( !findFoldersBelow(QDir(path), allSubfolders)) {
-        qDebug() << "Could not traverse all sub folders";
+int FolderWatcherPrivate::addFolderRecursiveHelper(const QString &path, const QSet<QString> &watchedFolders)
+{
+    QDir dir(path);
+    if( !(dir.exists() && dir.isReadable()) ) {
+        qDebug() << "Non existing path coming in: " << dir.absolutePath();
+        return 0;
     }
-    // qDebug() << "currently watching " << watchedFolders;
-    QStringListIterator subfoldersIt(allSubfolders);
-    while (subfoldersIt.hasNext()) {
-        QString subfolder = subfoldersIt.next();
-        // qDebug() << "  (**) subfolder: " << subfolder;
-        QDir folder (subfolder);
-        if (folder.exists() && !watchedFolders.contains(folder.absolutePath())) {
-            subdirs++;
-            if( _parent->pathIsIgnored(subfolder) ) {
+    int subdirs = 1;
+
+    inotifyRegisterPath(dir.absolutePath());
+
+    QDir::Filters filter = QDir::Dirs | QDir::NoDotAndDotDot | QDir::NoSymLinks | QDir::Hidden;
+    const QStringList pathes = dir.entryList(filter);
+    for (auto constIterator = pathes.constBegin(); constIterator != pathes.constEnd(); ++constIterator) {
+        const QString subfolder = path + QLatin1String("/") + (*constIterator);
+        QDir folder(subfolder);
+        if (folder.exists() && !watchedFolders.contains(subfolder)) {
+#ifndef OWNCLOUD_TEST // InotifyWatcherTest is not interested in ignored files and does not link against the folder
+            if( _parent->_folder->syncEngine().excludedFiles().isExcluded(
+                    subfolder, path, _parent->_folder->ignoreHiddenFiles())) {
```

Comment:
I think for correctness this must always pass in the full folder-relative path. (see other comment)